### DISCUSSION TITLE
feof error when uploading file with Symfony HTTP Client

### DISCUSSION THREAD

**Discussion by samhastings** [symfony/symfony#51360] [Q&A]
I'm using Symfony 6.3 with PHP 8.2 on macOS and trying to upload a file using Symfony's HTTP client:

```php
$file = fopen('/path/to/image.jpg', 'r');
$this->client->request('POST', '/image/processing/endpoint', [
    'body' => [
        'image_file' => $file,
    ],
]);
```

However, running this code is causing a `TransportException` with the message:

> Uploaded stream ended for body part "image_file".

The exception in question is being raised in line 384 of `Symfony\Component\HttpClient\HttpClientTrait` as follows:

```php
if (feof($v)) {
    throw new TransportException(sprintf('Uploaded stream ended for body part "%s".', $k));
}
```

It seems like the end-of-file test is returning `true` for some reason, however if I add the following line directly before I create the request in the code above, it is outputting `bool(false)`:

```php
$file = fopen('/path/to/image.jpg', 'r');
var_dump(feof($file));
// snip
```

Any other checks for file validity (such as `file_exists` and `is_readable` on the path) are confirming that the file path is definitely correct!

I am well and truly stumped, and wondered if anyone here could give me any advice on how to overcome this issue please?

**Comment by samhastings** [comment]
For anyone coming across this thread after Googling the same issue, I found a workaround.

Take a look at the [HTTP Client version 6.2 docs](https://symfony.com/doc/6.2/http_client.html#uploading-data) for an alternative (and slightly more long-winded) way to upload files inside a request body.

The relevant code snippet involves the use of the `DataPart` and `FormDataPart` classes from the Mime component.

Would still appreciate any input that sheds some light on the original issue!

**Comment by MatTheCat** [comment]
I managed to reproduce the issue: it comes from the `HttpClientDataCollector` which normalizes the request’s body again to generate the cURL command.

Opened #51366 and submitted #52472.
